lib/Sema/SemaExpr.cpp

```diff
@@ -12449,6 +12449,63 @@ ExprResult Sema::CreatePositionalParameterExpr(unsigned Index, QualType QT) {
   return new (Context) PositionalParameterExpr(Index, QT);
 }
 
+ExprResult Sema::ActOnBoundsCastExpr(Scope *S, SourceLocation LParenLoc,
+                                     Declarator &D, ParsedType &Ty,
+                                     SourceLocation RParenLoc, Expr *BaseExpr,
+                                     Expr *CountExpr, Expr *RangeExpr,
+                                     BoundsCastExpr::Kind kind) {
+  assert(!D.isInvalidType() && (BaseExpr != nullptr) &&
+         "ActOnBoundsCastExpr(): missing type or expr");
+
+  TypeSourceInfo *castTInfo = GetTypeForDeclaratorCast(D, BaseExpr->getType());
+
+  if (D.isInvalidType())
+    return ExprError();
+
+  ExprResult Res = CorrectDelayedTyposInExpr(BaseExpr);
+  if (!Res.isUsable())
+    return ExprError();
+  BaseExpr = Res.get();
+
+  checkUnusedDeclAttributes(D);
+
+  BoundsExpr *bounds = nullptr;
+  ExprResult Result(true);
+
+  ExprResult ResCount = CorrectDelayedTyposInExpr(CountExpr);
+  ExprResult ResRange = CorrectDelayedTyposInExpr(RangeExpr);
+
+  if (!ResCount.isUsable() && !ResRange.isUsable()) {
```

Comment:
I think you should have the caller build the bounds expression and pass it in as an argument, instead of building it here.